projects/core/koin-core/src/commonMain/kotlin/org/koin/core/instance/InstanceFactory.kt

```diff
@@ -28,8 +30,12 @@ import org.koin.mp.Lockable
 /**
  * Koin Instance Holder
  * create/get/release an instance of given definition
+ *
+ * Implements [Comparable] to provide factories soring by instance creation order
  */
-abstract class InstanceFactory<T>(val beanDefinition: BeanDefinition<T>) : Lockable() {
+abstract class InstanceFactory<T>(val beanDefinition: BeanDefinition<T>) : Lockable(), Comparable<InstanceFactory<*>> {
+    private val instanceCreationOrderSet = AtomicBoolean(false)
+    private var instanceCreationOrderPosition = Long.MAX_VALUE
```

Comment:
The current design ties a child instance _natural comparison_ to its insertion ordering within a parent collection. A `Comparable` is documented as a way to define a _class's natural order_, not their insertion order. See [Comparable][1] and [Natural Sort Order][2]

Instead of the child keeping track of the insertion order, what if the `InstanceRegistry` would handle the instances collection and ordering?

Here's a possible approach:

* The `InstanceRegistry` could use a `ConcurrentLinkedDeque` to store keys, ensuring thread-safe ordering of factory registrations.
 * The `InstanceRegistry` would then access the deque to order the factory instances map to close them in the correct order.

[1]: https://docs.oracle.com/en/java/javase/21/docs/api/java.base/java/lang/Comparable.html
[2]: https://en.m.wikipedia.org/wiki/Natural_sort_order

Reply:
same thought here. Consider another proposal @shchuko